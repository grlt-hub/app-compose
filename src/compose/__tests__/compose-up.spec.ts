import type { StageTuples } from '@prepareStages';
import { createRandomContainer } from '@randomContainer';
import { compose } from '../index';

const topology = createRandomContainer({ id: 'topology' });
const entityAccount = createRandomContainer({ id: 'entityAccount', dependencies: [topology] });
const entityLocale = createRandomContainer({ id: 'entityLocale', dependencies: [topology] });
const createFeatureAccountsList = (params?: Parameters<typeof createRandomContainer>[0]) =>
  createRandomContainer({ dependencies: [entityAccount], id: 'featureAccountsList', ...params });
const featureLanguageSelector = createRandomContainer({ id: 'featureLanguageSelector', dependencies: [entityLocale] });

describe('compose.up', () => {
  const featureAccountsList = createFeatureAccountsList();
  const featureDeposit = createRandomContainer({ dependencies: [featureAccountsList] });
  const featureSupportChat = createRandomContainer({
    dependencies: [entityLocale],
    optionalDependencies: [featureAccountsList],
  });

  test('required=all | success', async () => {
    const stages: StageTuples = [
      ['topology-stage', [topology]],
      ['entities-stage', [entityAccount, entityLocale]],
      ['first-order-features-stage', [featureAccountsList, featureDeposit]],
      ['second-order-features-stage', [featureLanguageSelector]],
      ['other-features-stage', [featureSupportChat]],
    ];

    const cmd = await compose({ stages, required: 'all' });
    const app = await cmd.up({ debug: false });

    expect(app.allDone).toBe(true);
    expect(app.stages).toStrictEqual({
      'topology-stage': { allDone: true, containerStatuses: { [topology.id]: 'done' } },
      'entities-stage': {
        allDone: true,
        containerStatuses: { [entityAccount.id]: 'done', [entityLocale.id]: 'done' },
      },
      'first-order-features-stage': {
        allDone: true,
        containerStatuses: {
          [featureAccountsList.id]: 'done',
          [featureDeposit.id]: 'done',
        },
      },
      'second-order-features-stage': {
        allDone: true,
        containerStatuses: { [featureLanguageSelector.id]: 'done' },
      },
      'other-features-stage': {
        allDone: true,
        containerStatuses: { [featureSupportChat.id]: 'done' },
      },
    });
  });

  test('required=all | fail', async () => {
    const featureAccountsList = createFeatureAccountsList({
      enable: () => false,
    });
    const featureDeposit = createRandomContainer({ id: 'featureDeposit', dependencies: [featureAccountsList] });
    const featureSupportChat = createRandomContainer({
      id: 'featureSupportChat',
      dependencies: [entityLocale],
      optionalDependencies: [featureAccountsList],
    });

    const stages: StageTuples = [
      ['topology-stage', [topology]],
      ['entities-stage', [entityAccount, entityLocale]],
      ['first-order-features-stage', [featureAccountsList, featureDeposit]],
      ['second-order-features-stage', [featureLanguageSelector]],
      ['other-features-stage', [featureSupportChat]],
    ];

    const cmd = await compose({ stages, required: 'all' });

    await expect(cmd.up()).rejects.toMatchSnapshot();
  });

  test('required=undefined | success', async () => {
    const featureAccountsList = createFeatureAccountsList({
      enable: () => false,
    });
    const featureDeposit = createRandomContainer({ id: 'featureDeposit', dependencies: [featureAccountsList] });
    const featureSupportChat = createRandomContainer({
      id: 'featureSupportChat',
      dependencies: [entityLocale],
      optionalDependencies: [featureAccountsList],
    });

    const stages: StageTuples = [
      ['topology-stage', [topology]],
      ['entities-stage', [entityAccount, entityLocale]],
      ['first-order-features-stage', [featureAccountsList, featureDeposit]],
      ['second-order-features-stage', [featureLanguageSelector]],
      ['other-features-stage', [featureSupportChat]],
    ];

    const cmd = await compose({ stages });
    const app = await cmd.up();

    expect(app.allDone).toBe(false);
    expect(app.stages).toStrictEqual({
      'topology-stage': { allDone: true, containerStatuses: { [topology.id]: 'done' } },
      'entities-stage': {
        allDone: true,
        containerStatuses: { [entityAccount.id]: 'done', [entityLocale.id]: 'done' },
      },
      'first-order-features-stage': {
        allDone: false,
        containerStatuses: {
          [featureAccountsList.id]: 'off',
          [featureDeposit.id]: 'off',
        },
      },
      'second-order-features-stage': {
        allDone: true,
        containerStatuses: { [featureLanguageSelector.id]: 'done' },
      },
      'other-features-stage': {
        allDone: true,
        containerStatuses: { [featureSupportChat.id]: 'done' },
      },
    });
  });

  test('required=list | success', async () => {
    const featureAccountsList = createFeatureAccountsList({
      enable: () => false,
    });
    const featureDeposit = createRandomContainer({ id: 'featureDeposit', dependencies: [featureAccountsList] });
    const featureSupportChat = createRandomContainer({
      id: 'featureSupportChat',
      dependencies: [entityLocale],
      optionalDependencies: [featureAccountsList],
    });

    const stages: StageTuples = [
      ['topology-stage', [topology]],
      ['entities-stage', [entityAccount, entityLocale]],
      ['first-order-features-stage', [featureAccountsList, featureDeposit]],
      ['second-order-features-stage', [featureLanguageSelector]],
      ['other-features-stage', [featureSupportChat]],
    ];

    const cmd = await compose({ stages, required: [topology, [featureLanguageSelector, featureAccountsList]] });
    const app = await cmd.up();

    expect(app.allDone).toBe(false);
    expect(app.stages).toStrictEqual({
      'topology-stage': { allDone: true, containerStatuses: { [topology.id]: 'done' } },
      'entities-stage': {
        allDone: true,
        containerStatuses: { [entityAccount.id]: 'done', [entityLocale.id]: 'done' },
      },
      'first-order-features-stage': {
        allDone: false,
        containerStatuses: {
          [featureAccountsList.id]: 'off',
          [featureDeposit.id]: 'off',
        },
      },
      'second-order-features-stage': {
        allDone: true,
        containerStatuses: { [featureLanguageSelector.id]: 'done' },
      },
      'other-features-stage': {
        allDone: true,
        containerStatuses: { [featureSupportChat.id]: 'done' },
      },
    });
  });

  test('required=list | fail | enable: () => false', async () => {
    const featureAccountsList = createFeatureAccountsList({ enable: () => false });
    const featureDeposit = createRandomContainer({ id: 'featureDeposit', dependencies: [featureAccountsList] });
    const featureSupportChat = createRandomContainer({
      id: 'featureSupportChat',
      dependencies: [entityLocale],
      optionalDependencies: [featureAccountsList],
    });

    const stages: StageTuples = [
      ['topology-stage', [topology]],
      ['entities-stage', [entityAccount, entityLocale]],
      ['first-order-features-stage', [featureAccountsList, featureDeposit]],
      ['second-order-features-stage', [featureLanguageSelector]],
      ['other-features-stage', [featureSupportChat]],
    ];

    const cmd = await compose({ stages, required: [topology, featureDeposit] });

    await expect(cmd.up()).rejects.toMatchSnapshot();
  });

  test('required=list | fail | throw new Error("oops")', async () => {
    const featureAccountsList = createFeatureAccountsList({
      enable: () => {
        throw new Error('oops');
      },
    });
    const featureDeposit = createRandomContainer({ id: 'featureDeposit', dependencies: [featureAccountsList] });
    const featureSupportChat = createRandomContainer({
      id: 'featureSupportChat',
      dependencies: [entityLocale],
      optionalDependencies: [featureAccountsList],
    });

    const stages: StageTuples = [
      ['topology-stage', [topology]],
      ['entities-stage', [entityAccount, entityLocale]],
      ['first-order-features-stage', [featureAccountsList, featureDeposit]],
      ['second-order-features-stage', [featureLanguageSelector]],
      ['other-features-stage', [featureSupportChat]],
    ];

    const cmd = await compose({ stages, required: [topology, featureDeposit] });

    await expect(cmd.up()).rejects.toMatchSnapshot();
  });

  test('required=list | success | one of', async () => {
    const featureAccountsList = createFeatureAccountsList({
      enable: () => {
        throw new Error('oops');
      },
    });
    const featureDeposit = createRandomContainer({ id: 'featureDeposit', dependencies: [featureAccountsList] });
    const featureSupportChat = createRandomContainer({
      id: 'featureSupportChat',
      dependencies: [entityLocale],
      optionalDependencies: [featureAccountsList],
    });

    const stages: StageTuples = [
      ['topology-stage', [topology]],
      ['entities-stage', [entityAccount, entityLocale]],
      ['first-order-features-stage', [featureAccountsList, featureDeposit]],
      ['second-order-features-stage', [featureLanguageSelector]],
      ['other-features-stage', [featureSupportChat]],
    ];

    const cmd = await compose({ stages, required: [topology, [featureDeposit, featureSupportChat]] });
    const app = await cmd.up();

    expect(app.allDone).toBe(false);
    expect(app.stages).toStrictEqual({
      'topology-stage': { allDone: true, containerStatuses: { [topology.id]: 'done' } },
      'entities-stage': {
        allDone: true,
        containerStatuses: { [entityAccount.id]: 'done', [entityLocale.id]: 'done' },
      },
      'first-order-features-stage': {
        allDone: false,
        containerStatuses: {
          [featureAccountsList.id]: 'fail',
          [featureDeposit.id]: 'fail',
        },
      },
      'second-order-features-stage': {
        allDone: true,
        containerStatuses: { [featureLanguageSelector.id]: 'done' },
      },
      'other-features-stage': {
        allDone: true,
        containerStatuses: { [featureSupportChat.id]: 'done' },
      },
    });
  });
});
