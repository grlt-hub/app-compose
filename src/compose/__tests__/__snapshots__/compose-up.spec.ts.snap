// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compose.up > required=all | fail 1`] = `
[Error: [app-compose] Application startup failed.
Required container(s) "featureAccountsList" did not up in stage "first-order-features-stage".

Startup Log:

  topology-stage: 
    allDone: true
    containerStatuses: 
      topology: done

  entities-stage: 
    allDone: true
    containerStatuses: 
      entityAccount: done
      entityLocale: done

  first-order-features-stage: 
    allDone: false
    containerStatuses: 
      featureAccountsList: off
      featureDeposit: off


Recommendations:
- Verify if the container(s) "featureAccountsList" are truly required.
- If not, consider removing them from the required list in "up.required".
- Ensure all dependencies for the container(s) are correct and their logic works as expected.]
`;

exports[`compose.up > required=list | fail | enable: () => false 1`] = `
[Error: [app-compose] Application startup failed.
Required container(s) "featureDeposit" did not up in stage "first-order-features-stage".

Startup Log:

  topology-stage: 
    allDone: true
    containerStatuses: 
      topology: done

  entities-stage: 
    allDone: true
    containerStatuses: 
      entityAccount: done
      entityLocale: done

  first-order-features-stage: 
    allDone: false
    containerStatuses: 
      featureAccountsList: off
      featureDeposit: off


Recommendations:
- Verify if the container(s) "featureDeposit" are truly required.
- If not, consider removing them from the required list in "up.required".
- Ensure all dependencies for the container(s) are correct and their logic works as expected.]
`;

exports[`compose.up > required=list | fail | throw new Error("oops") 1`] = `
[Error: [app-compose] Application startup failed.
Required container(s) "featureDeposit" did not up in stage "first-order-features-stage".

Startup Log:

  topology-stage: 
    allDone: true
    containerStatuses: 
      topology: done

  entities-stage: 
    allDone: true
    containerStatuses: 
      entityAccount: done
      entityLocale: done

  first-order-features-stage: 
    allDone: false
    containerStatuses: 
      featureAccountsList: fail
      featureDeposit: fail


Recommendations:
- Verify if the container(s) "featureDeposit" are truly required.
- If not, consider removing them from the required list in "up.required".
- Ensure all dependencies for the container(s) are correct and their logic works as expected.]
`;
